plugin/code_actions.py

```diff
@@ -12,37 +12,64 @@
 from .core.diagnostics import get_point_diagnostics
 from .core.url import filename_to_uri
 from .core.views import region_to_range
-from .core.helpers import debounce
 from .core.registry import session_for_view
 from .core.settings import settings
 
 
-class CodeAction:
-    def __init__(self, view: 'sublime.View') -> None:
-        self.view = view
+def send_code_action_request(view, on_response_recieved: 'Callable'):
+    session = session_for_view(view)
+    if not session or not session.has_capability('codeActionProvider'):
+        # the server doesn't support code actions, just return
+        return
+
+    region = view.sel()[0]
+    pos = region.begin()
+    point_diagnostics = get_point_diagnostics(view, pos)
+    params = {
+        "textDocument": {
+            "uri": filename_to_uri(view.file_name())
+        },
+        "range": region_to_range(view, region).to_lsp(),
+        "context": {
+            "diagnostics": list(diagnostic.to_lsp() for diagnostic in point_diagnostics)
+        }
+    }
+    session.client.send_request(
+        Request.codeAction(params),
+        lambda response: on_response_recieved(response))
 
-    def send_request(self, on_response_recieved: 'Optional[Callable]' = None):
-        """ callback - hook with response as the first argument. """
-        session = session_for_view(self.view)
-        if not session or not session.has_capability('codeActionProvider'):
-            # the server doesn't support code actions, just return
-            return
 
-        params = self._get_code_action_params()
-        session.client.send_request(
-            Request.codeAction(params),
-            lambda response: self._handle_response(response, on_response_recieved))
+class LspCodeActionBulbListener(sublime_plugin.ViewEventListener):
+    def __init__(self, view: sublime.View) -> None:
+        super().__init__(view)
+        self._stored_point = -1
 
-    def _handle_response(self, response, callback: 'Optional[Callable]' = None) -> None:
-        code_action = CodeAction(self.view)
+    @classmethod
+    def is_applicable(cls, _settings):
+        if settings.show_code_actions_bulb:
+            return True
+        return False
+
+    def on_selection_modified_async(self):
+        self.hide_bulb()
+        self.fire_request()
+
+    def fire_request(self):
+        current_point = self.view.sel()[0].begin()
+        if self._stored_point != current_point:
+            self._stored_point = current_point
+            sublime.set_timeout_async(lambda: self._purge(current_point), 800)
+
+    def _purge(self, current_point: int) -> None:
```

Comment:
I used the word `purge` for the document sync logic as it empties a queue of updates - here we only have a single value instead of a list. Perhaps we can call these two `schedule_request` and `fire_request` ?